operators/azure/curl_invoker.hpp

```diff
@@ -0,0 +1,98 @@
+
+// Licensed under the MIT License.
+#pragma once
+
+#include <memory>
+
+#include "curl/curl.h"
+
+#include "ocos.h"
+#include "cloud_base_kernel.hpp"
+
+namespace ort_extensions {
+
+class CurlHandler {
+ public:
+  using WriteCallBack = size_t (*)(char*, size_t, size_t, void*);
+
+  CurlHandler(WriteCallBack callback);
+  ~CurlHandler() = default;
+
+  /// <summary>
+  /// Callback to add contents to a string
+  /// </summary>
+  /// <seealso cref="https://curl.se/libcurl/c/CURLOPT_WRITEFUNCTION.html"/>
+  /// <returns>Bytes processed. If this does not match element_size * num_elements the libcurl function
+  /// used will return CURLE_WRITE_ERROR</returns>
+  static size_t WriteStringCallback(char* contents, size_t element_size, size_t num_elements, void* userdata);
+
+  void AddHeader(const char* data) {
+    headers_.reset(curl_slist_append(headers_.release(), data));
+  }
+
+  template <typename... Args>
+  void AddForm(Args... args) {
+    curl_formadd(&from_, &last_, args...);
+  }
+
+  void AddFormString(const char* name, const char* value) {
+    AddForm(CURLFORM_COPYNAME, name,
+            CURLFORM_COPYCONTENTS, value,
+            CURLFORM_END);
+  }
+
+  void AddFormBuffer(const char* name, const char* buffer_name, const void* buffer_ptr, size_t buffer_len) {
+    AddForm(CURLFORM_COPYNAME, name,
+            CURLFORM_BUFFER, buffer_name,
+            CURLFORM_BUFFERPTR, buffer_ptr,
+            CURLFORM_BUFFERLENGTH, buffer_len,
+            CURLFORM_END);
+  }
+
+  template <typename T>
+  void SetOption(CURLoption opt, T val) {
+    curl_easy_setopt(curl_.get(), opt, val);
+  }
+
+  CURLcode Perform() {
+    SetOption(CURLOPT_HTTPHEADER, headers_.get());
+    if (from_) {
+      SetOption(CURLOPT_HTTPPOST, from_);
+    }
+
+    return curl_easy_perform(curl_.get());
+  }
+
+ private:
+  std::unique_ptr<CURL, decltype(curl_easy_cleanup)*> curl_;
+  std::unique_ptr<curl_slist, decltype(curl_slist_free_all)*> headers_;
+  curl_httppost* from_{};
+  curl_httppost* last_{};
+  std::unique_ptr<curl_httppost, decltype(curl_formfree)*> from_holder_;  // TODO: Why no last_holder_?
+};
+
+/// <summary>
+/// Base class for requests using Curl
+/// </summary>
+class CurlInvoker : public CloudBaseKernel {
+ protected:
+  CurlInvoker(const OrtApi& api, const OrtKernelInfo& info);
+  virtual ~CurlInvoker() = default;
+
+ public:
+  // Compute method that is used to co-ordinate all Curl based Azure requests
+  // TODO: Can this be `const` to enforce it's stateless?
```

Comment:
Yes, all kernels should be stateless. Hence this must be a const method.

Reply:
https://github.com/microsoft/onnxruntime-extensions/pull/500

Will update the kernels here once that is checked in as the OrtLiteCustomStruct needs to be changed first.